### PR TITLE
Adding Lag Features for localmode

### PR DESCRIPTION
**Pull request by ahmadnazeri**
# Description
This PR adds lag features for localmode. 

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: client/src/featureform/sqlite_metadata.py [hunk 1]
```diff
@@ -97,7 +97,9 @@ def createTables(self):
           training_set_variant text NOT NULL,
           feature_name text NOT NULL,
           feature_variant text NOT NULL,
-          UNIQUE(training_set_name, training_set_variant, feature_name, feature_variant));''')
+          feature_new_name text,
+          feature_lag real,
+          UNIQUE(training_set_name, training_set_variant, feature_name, feature_variant, feature_new_name, feature_lag));''')
```
Comment (sdreyer) [inline client/src/featureform/sqlite_metadata.py:102]: Only concern is i dont believe this is backwards compatible

File: client/src/featureform/resources.py [hunk 2]
```diff
@@ -998,19 +998,43 @@ def _check_insert_training_set_resources(self, db) -> None:
...
 
+        for feature in self.feature_lags:
+            feature_name = feature["feature"]
+            feature_variant = feature["variant"]
+            feature_new_name = feature["name"]
+            feature_lag = feature["lag"].total_seconds()
+
+            try:
+                db.get_feature_variant(feature_name, feature_variant)
+            except ValueError:
```
Comment (Sami1309) [inline client/src/featureform/resources.py:1024]: perhaps make it ValueError as err and add err to the returned error

**Review by sdreyer — DISMISSED** [review]

File: provider/scripts/k8s/tests/conftest.py [hunk 3]
```diff
@@ -140,10 +140,15 @@ def transformation(transaction):
...
 @pytest.fixture(scope="module")
 def container_client():
-    # get the path to .env in root directory
-    env_file = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(real_path)))))
-    load_dotenv(f"{env_file}/.env")
-    blob_service_client = BlobServiceClient.from_connection_string(os.getenv("AZURE_CONNECTION_STRING"))
+    connection_string = os.getenv("AZURE_CONNECTION_STRING")
+    if connection_string == None:
+        # get the path to .env in root directory
+        env_file = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(real_path)))))
```
Comment (Sami1309) [inline provider/scripts/k8s/tests/conftest.py:146]: what ... is going on here

File: provider/scripts/k8s/offline_store_pandas_runner.py [hunk 4]
```diff
@@ -159,10 +164,10 @@ def download_blobs_to_local(container_client, blob, local_filename):
     
     print(f"downloading {blob} to {local_filename}")
     if not os.path.isdir(LOCAL_DATA_PATH):
-        os.makedirs(LOCAL_DATA_PATH)
+        os.makedirs(LOCAL_DATA_PATH, exist_ok=True)
 
     full_path = f"{LOCAL_DATA_PATH}/{local_filename}"
-    if blob[-4:] == ".csv" or blob[-8:] == ".parquet":
+    if blob[-4:] == ".csv" or blob[-8:] == ".parquet" or blob[-4:] == ".pkl":
```
Comment (Sami1309) [inline provider/scripts/k8s/offline_store_pandas_runner.py:170]: might be easier to split on the final "." and compare that to a set of valid extensions

**Review by Sami1309 — APPROVED** [review]
lg, just a couple minor comments and a question of the nested dir functions

File: provider/scripts/k8s/tests/conftest.py (hunk 3 above)
Comment (sdreyer) [reply provider/scripts/k8s/tests/conftest.py:146]: +1

**Pull request merged**